In [1]:
import turicreate as tc
import pandas as pd

### I removed these words; and, of , it, the.  These stop words have very little useful information that we need.

In [2]:
movies = tc.SFrame('./IMDB_Dataset_mod.csv')
movies

Finished parsing file /Users/ola/SupervisedML/Assignment4/sentiment_analysis/IMDB_Dataset_mod.csv

Parsing completed. Parsed 100 lines in 0.785446 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 42865 lines. Lines per second: 35670

Finished parsing file /Users/ola/SupervisedML/Assignment4/sentiment_analysis/IMDB_Dataset_mod.csv

Parsing completed. Parsed 50000 lines in 1.35147 secs.

Columns:
	review	str
	sentiment	str

Rows: 50000

Data:
+-------------------------------+-----------+
|             review            | sentiment |
+-------------------------------+-----------+
| One other reviewers has me... |  positive |
| A wonderful little product... |  positive |
| I thought this was a wonde... |  positive |
| Basically there's a family... |  negative |
| Petter Mattei's "Love in T... |  positive |
| Probably my all-time favor... |  positive |
| I sure would like to see a... |  positive |
| This show was an amazing, ... |  negative |
| Encouraged by positive com... |  negative |
| If you like original gut w... |  positive |
+-------------------------------+-----------+
[50000 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [3]:
movies['words'] = tc.text_analytics.count_words(movies['review'])
movies

Columns:
	review	str
	sentiment	str
	words	dict

Rows: 50000

Data:
+-------------------------------+-----------+-------------------------------+
|             review            | sentiment |             words             |
+-------------------------------+-----------+-------------------------------+
| One other reviewers has me... |  positive | {'if': 1.0, 'viewing': 1.0... |
| A wonderful little product... |  positive | {'done': 1.0, 'every': 1.0... |
| I thought this was a wonde... |  positive | {'see': 1.0, 'go': 1.0, 'g... |
| Basically there's a family... |  negative | {'them': 1.0, 'ignore': 1.... |
| Petter Mattei's "Love in T... |  positive | {'work': 1.0, 'his': 1.0, ... |
| Probably my all-time favor... |  positive | {'for': 1.0, 'dozen': 1.0,... |
| I sure would like to see a... |  positive | {'it': 1.0, 'do': 1.0, 'th... |
| This show was an amazing, ... |  negative | {'awful': 1.0, 'just': 1.0... |
| Encouraged by positive com... |  negative | {'effort': 1.0, 'an': 1.0,... |
| If you like original gut w... |  positive | {'camp': 1.0, 'great': 1.0... |
+-------------------------------+-----------+-------------------------------+
[50000 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [4]:
train, test = movies.random_split(.80, seed=0)

In [5]:
model = tc.logistic_classifier.create(train, features=['words'], target='sentiment', l1_penalty=0.001, l2_penalty=0.00001)

PROGRESS: Creating a validation set from 5 percent of training data. This may take a while.
          You can set ``validation_set=None`` to disable validation tracking.



Logistic regression:

--------------------------------------------------------

Number of examples          : 38030

Number of classes           : 2

Number of feature columns   : 1

Number of unpacked features : 92421

Number of coefficients      : 92422

Starting Accelerated Gradient (FISTA)

--------------------------------------------------------

Tuning step size. First iteration could take longer than subsequent iterations.

+-----------+----------+-----------+--------------+-------------------+---------------------+

| Iteration | Passes   | Step size | Elapsed Time | Training Accuracy | Validation Accuracy |

+-----------+----------+-----------+--------------+-------------------+---------------------+

| 0         | 1        | 1.000000  | 1.279056     | 0.500158          | 0.514985            |

| 1         | 2        | 0.000008  | 7.223521     | 0.922535          | 0.852647            |

| 2         | 3        | 0.000008  | 7.743613     | 0.962714          | 0.877123            |

| 3         | 4        | 0.000008  | 8.267606     | 0.975072          | 0.889610            |

| 4         | 5        | 0.000008  | 8.802615     | 0.980752          | 0.895105            |

| 5         | 6        | 0.000008  | 9.336122     | 0.983355          | 0.893606            |

| 9         | 10       | 0.000005  | 11.800479    | 0.989955          | 0.892607            |

+-----------+----------+-----------+--------------+-------------------+---------------------+

SUCCESS: Optimal solution found.

##### Training Accuracy = 0.98 and Validation Accuracy = 0.90

In [6]:
model_predictions = test['review', 'sentiment']
model_predictions['model_predictions'] = model.predict(test)
model_predictions

Columns:
	review	str
	sentiment	str
	model_predictions	str

Rows: 9968

Data:
+-------------------------------+-----------+-------------------+
|             review            | sentiment | model_predictions |
+-------------------------------+-----------+-------------------+
| So im not a big fan Boll's... |  negative |      negative     |
| Kind drawn in by erotic sc... |  negative |      negative     |
| The Karen Carpenter Story ... |  positive |      positive     |
| Taut organically gripping,... |  positive |      positive     |
| One most significant quote... |  positive |      positive     |
| After sitting through this... |  negative |      negative     |
| This movie is based on boo... |  positive |      positive     |
| The Hills Have Eyes II is ... |  negative |      negative     |
| I've just watched Fingersm... |  negative |      negative     |
| I thought that Mukhsin has... |  positive |      negative     |
+-------------------------------+-----------+-------------------+
[9968 rows x 3 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [7]:
testing_accuracy = tc.evaluation.accuracy(model_predictions['sentiment'], model_predictions['model_predictions'])
testing_accuracy

0.8942616372391653

2(g) Testing accuracy = 0.89

| l1_penalty |  l2_penalty | Training Accuracy | Validation Accuracy | Testing Accuracy |
| -: | -: | :-: | -: | :-: |
| 0.00 | 0.01 | 0.999711  | 0.874625 | 0.870886 |
| 0.00 | 0.001 | 0.999816 | 0.866633 | 0.871890 |
| 0.01 | 0.01 | 0.970313 | 0.882617 | 0.886235 |
| 0.01 | 0.001 | 0.968709 | 0.890609 | 0.888342 |
| 0.001 | 0.00001 | 0.981015 | 0.906593 | 0.894462 |



#### Optimal choice of L1 and L2 = 0.001 and 0.00001 because these regularization values help trained a model with higher accuracy for the validation set and testing set while training accuracy is good.